sklearn/inspection/_base.py

```diff
@@ -0,0 +1,365 @@
+        X = check_array(X, force_all_finite='allow-nan', dtype=np.object)
+
+    accepted_responses = ('auto', 'predict_proba', 'decision_function')
+    if response_method not in accepted_responses:
+        raise ValueError(
+            'response_method {} is invalid. Accepted response_method names '
+            'are {}.'.format(response_method, ', '.join(accepted_responses)))
+
+    if is_regressor(estimator) and response_method != 'auto':
+        raise ValueError(
+            "The response_method parameter is ignored for regressors and "
+            "must be 'auto'."
+        )
+
+    accepted_methods = ('brute', 'recursion', 'auto')
+    if method not in accepted_methods:
+        raise ValueError(
+            'method {} is invalid. Accepted method names are {}.'.format(
+                method, ', '.join(accepted_methods)))
+
+    if method == 'auto':
+        if (isinstance(estimator, BaseGradientBoosting) and
+                estimator.init is None):
+            method = 'recursion'
+        elif isinstance(estimator, BaseHistGradientBoosting):
+            method = 'recursion'
+        else:
+            method = 'brute'
+
+    if method == 'recursion':
+        if not isinstance(estimator,
+                          (BaseGradientBoosting, BaseHistGradientBoosting)):
+            supported_classes_recursion = (
+                'GradientBoostingClassifier',
+                'GradientBoostingRegressor',
+                'HistGradientBoostingClassifier',
+                'HistGradientBoostingRegressor',
+            )
+            raise ValueError(
+                "Only the following estimators support the 'recursion' "
+                "method: {}. Try using method='brute'."
+                .format(', '.join(supported_classes_recursion)))
+        if response_method == 'auto':
+            response_method = 'decision_function'
+
+        if response_method != 'decision_function':
+            raise ValueError(
+                "With the 'recursion' method, the response_method must be "
+                "'decision_function'. Got {}.".format(response_method)
+            )
+
+    if _determine_key_type(features, accept_slice=False) == 'int':
+        # _get_column_indices() supports negative indexing. Here, we limit
+        # the indexing to be positive. The upper bound will be checked
+        # by _get_column_indices()
+        if np.any(np.less(features, 0)):
+            raise ValueError(
+                'all features must be in [0, {}]'.format(X.shape[1] - 1)
+            )
+
+    return X, response_method, method
+
+
+def _plot_pdp_ice(estimator, X, features, calc_method, display_class,
+                  feature_names=None, target=None, response_method='auto',
+                  n_cols=3, grid_resolution=100, percentiles=(0.05, 0.95),
+                  method='auto', n_jobs=None, verbose=0, fig=None,
+                  line_kw=None, contour_kw=None, ax=None, **kwargs):
+
```

Comment:
it might be helpful to have a docstring that describes what the role of this helper is, and particularly they key params that differentiate between ICE and PDP

Reply:
Thanks for the suggestion. Will add that.

Reply:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/feae02c7d2cc94ea10a2fd96b8a2a145bf14c891